src/utils/net/download.rs

```diff
@@ -1,202 +1,91 @@
+            _ => {
+                info!("reading file: {}", path);
+                let stream = tokio::fs::File::open(path).await?;
+                let content_length = stream.metadata().await?.len();
+                (Right(stream), content_length)
+            }
+        };
+
+        if content_length > 0 {
+            read_progress.set_length(content_length);
+            read_progress.set_style(Self::progress_style())
         }
-        r
-    }
-}
-
-type DownloadStream = IntoAsyncRead<MapErr<hyper::Body, fn(hyper::Error) -> futures::io::Error>>;
-
-impl FetchProgress<DownloadStream> {
-    pub async fn fetch_from_url(url: &Url) -> anyhow::Result<Self> {
-        let (inner, progress_bar) = fetch_stream_from_url(url).await?;
-        Ok(FetchProgress {
-            inner,
-            progress_bar,
-        })
-    }
-}
-
-impl FetchProgress<ZstdDecoder<DownloadStream>> {
-    pub async fn fetch_zstd_compressed_from_url(url: &Url) -> anyhow::Result<Self> {
-        let (inner, progress_bar) = fetch_stream_from_url(url).await?;
-        let inner = ZstdDecoder::new(inner);
-        Ok(FetchProgress {
-            inner,
-            progress_bar,
-        })
-    }
-}
-
-impl FetchProgress<BufReader<async_fs::File>> {
-    async fn fetch_from_file(file: async_fs::File) -> anyhow::Result<Self> {
-        let total_size = file.metadata().await?.len();
 
-        let pb = ProgressBar::new(total_size);
-        pb.message("Importing snapshot ");
-        pb.set_units(crate::utils::io::progress_bar::Units::Bytes);
-        pb.set_max_refresh_rate(Some(Duration::from_millis(500)));
+        let mut reader = tokio::io::BufReader::new(read_progress.wrap_async_read(stream));
 
-        Ok(FetchProgress {
-            progress_bar: pb,
-            inner: BufReader::new(file),
-        })
+        Ok(Box::new(
+            match Self::is_zstd(reader.fill_buf().await?).await? {
+                true => Left(ZstdDecoder::new(reader)),
+                false => Right(reader),
+            }
+            .compat(),
+        ))
     }
-}
-
-impl FetchProgress<ZstdDecoder<BufReader<async_fs::File>>> {
-    async fn fetch_from_zstd_compressed_file(file: async_fs::File) -> anyhow::Result<Self> {
-        let total_size = file.metadata().await?.len();
-
-        let pb = ProgressBar::new(total_size);
-        pb.message("Importing snapshot ");
-        pb.set_units(crate::utils::io::progress_bar::Units::Bytes);
-        pb.set_max_refresh_rate(Some(Duration::from_millis(500)));
-
-        let inner = ZstdDecoder::new(BufReader::new(file));
 
-        Ok(FetchProgress {
-            progress_bar: pb,
-            inner,
-        })
+    // This method checks the header in order to see whether or not we are operating on a zstd
+    // archive. The zstd header has a maximum size of 18 bytes:
+    // https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#zstandard-frames.
+    async fn is_zstd(buf: &[u8]) -> anyhow::Result<bool> {
```

Comment:
This function looks to be neither async nor fallible.

```suggestion
    fn is_zstd(buf: &[u8]) -> bool {
```

Reply:
Yes, I've made changes to it, but forgot to change the signature. Good catch, thanks!